### PR TITLE
[Graophbolt] Add utils for sample

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add utility functions for sampling.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    compacted_nodes_dict = {}
+    for ntype, nodes in nodes_dict.items():
+        collected_nodes = torch.cat(nodes)
+        # Compact and find unique nodes.
+        unique_nodes, collected_nodes = torch.unique(
+            collected_nodes,
+            return_inverse=True,
+        )
+        unique_nodes_dict[ntype] = unique_nodes
+        compacted_nodes_dict[ntype] = collected_nodes
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:56]:
```suggestion
        inverse_indices_dict[ntype] = inverse_indices
```

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        u, v = node_pair
+        nodes_dict[u_type].append(u)
+        nodes_dict[v_type].append(v)
+
+    unique_nodes_dict = {}
+    compacted_nodes_dict = {}
+    for ntype, nodes in nodes_dict.items():
+        collected_nodes = torch.cat(nodes)
+        # Compact and find unique nodes.
+        unique_nodes, collected_nodes = torch.unique(
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:51]: change for fix

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        Also, it should be in the format ('u', 'v') representing source
+        and destination pairs. And IDs inside are homogeneous ids.
+        - If `node_pairs` is a dictionary: The keys should be edge type and
+        the values should be corresponding node pairs. And IDs inside are
+        heterogeneous ids.
+    
+    Returns
+        -------
+        Tuple[node_pairs, unique_nodes]
+            The compacted node pairs and the unique nodes (per type).
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:33]: added

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 1 above)
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:56]: changed for fix

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+
+def unique_and_compact_node_pairs(
+    node_pairs: Union[
+        Tuple[torch.Tensor, torch.Tensor],
+        Dict[Tuple[str, str, str], Tuple[torch.Tensor, torch.Tensor]],
+    ]
+):
+    """
+    Compact node pairs and return unique nodes (per type).
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:16]: added

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        pairs are replaced with mapped node IDs, where each type of node is
+        mapped to a contiguous space of IDs ranging from 0 to N.
+
+    Examples
+    --------
+    >>> import dgl.graphbolt as gb
+    >>> N1 = torch.LongTensor([1, 2, 2])
+    >>> N2 = torch.LongTensor([5, 6, 5])
+    >>> node_pairs = {("n1", "e1", "n2"): (N1, N2),
+    ...  ("n2", "e2", "n1"): (N2, N1)}
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:45]:
```suggestion
    ...     ("n2", "e2", "n1"): (N2, N1)}
```

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    >>> unique_nodes, compacted_node_pairs = gb.unique_and_compact_node_pairs(
+    ... node_pairs
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:47]:
```suggestion
    ...     node_pairs
```

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    for etype, node_pair in node_pairs.items():
+        u_type, _, v_type = etype
+        u, v = node_pair
+        u_size, v_size = u.numel(), v.numel()
+        u = inverse_indices_dict[u_type][:u_size]
+        inverse_indices_dict[u_type] = inverse_indices_dict[u_type][u_size:]
+        v = inverse_indices_dict[v_type][:v_size]
+        inverse_indices_dict[v_type] = inverse_indices_dict[v_type][v_size:]
+        compacted_node_pairs[etype] = (u, v)
+    if not is_heterogeneous:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:90]:
```suggestion

    # Return singleton for homogeneous graph.
    if not is_heterogeneous:
```

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    >>> unique_nodes, compacted_node_pairs = gb.unique_and_compact_node_pairs(
+    ... node_pairs
+    ... )
+    >>> print(unique_nodes)
+    {'n1': tensor([1, 2]), 'n2': tensor([5, 6])}
+    >>> print(compacted_node_pairs)
+    {('n1', 'e1', 'n2'): (tensor([0, 1, 1]), tensor([0, 1, 0])),
+    ('n2', 'e2', 'n1'): (tensor([0, 1, 0]), tensor([0, 1, 1]))}
+    """
+    is_heterogeneous = isinstance(node_pairs, Dict)
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:55]: Such that the 2 `if not is_heterogeneous` can be replaced with `if is_homogeneous`

```suggestion
    is_homogeneous = not isinstance(node_pairs, Dict)
```

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 5 above)
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:45]: fixed

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 6 above)
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:47]: fixed

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 8 above)
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:55]: make sense, use `is_homogeneous `

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 7 above)
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:90]: use `is_homogeneous `

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    for etype, node_pair in node_pairs.items():
+        u_type, _, v_type = etype
+        u, v = node_pair
+        u_size, v_size = u.numel(), v.numel()
+        u = inverse_indices_dict[u_type][:u_size]
+        inverse_indices_dict[u_type] = inverse_indices_dict[u_type][u_size:]
+        v = inverse_indices_dict[v_type][:v_size]
+        inverse_indices_dict[v_type] = inverse_indices_dict[v_type][v_size:]
+        compacted_node_pairs[etype] = (u, v)
+    if is_homogeneous:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:90]:
```suggestion

    # Return singleton for homogeneous graph.
    if is_homogeneous:
```

**Pull request merged**